### PR TITLE
Add time and energy measurements

### PR DESCRIPTION
**Pull request by OsmanSeckinSimsek**
This commit adds time and energy measurements for each function.
Energy is measured using PMT on Cray systems if enabled.

### CODE REVIEW

File: main/src/util/profiler.hpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+    void printProfilingInfoHDF5(size_t timesteps)
+    {
+        H5PartSetStep(_profilingFile, 0);
+        H5PartSetNumParticles(_profilingFile, funcTimes.size());
+        int numFuncs = funcTimes.size() / timesteps;
+        fileutils::writeH5PartStepAttrib(_profilingFile, "NUM_RANKS", &_numRanks, 1);
+        fileutils::writeH5PartStepAttrib(_profilingFile, "NUM_FUNCS", &numFuncs, 1);
+        std::string fieldName = "RANK" + std::to_string(_rank);
+        fileutils::writeH5PartField(_profilingFile, fieldName, funcTimes.data());
```
Comment (sekelle) [inline main/src/util/profiler.hpp]: This will give you a random label. For parallel writes, the field name has to match across ranks.

**Review by sekelle — CHANGES_REQUESTED** [review]
Thanks for the code. The biggest opportunity to improve this PR is by reusing the `IFileWriter` interface. This will a) get rid of separate HDF5 and Ascii write functions and b) avoid MPI gather code because `IFileWriter` takes care of that. This would be good news for modularization to cut down on compile times as well because we can compile the IO stack in separate unit.
A second factor to consider is to perform the energy measurements through an abstract interface. You shouldn't have to #ifdef every single function in the CrayPM reader and you might at some point want to get readings from a different provider.

File: main/src/util/profiler.hpp (hunk 1 above)
Comment (OsmanSeckinSimsek) [reply main/src/util/profiler.hpp]: I thought HDF5 could write into multiple labels in parallel. Is it not the case?

**Comment by sekelle** [comment]
I have reimplemented this for the following reasons:
* Counter reading (`pmt.h`) was missing
* Perform IO only through the abstract interface to eliminate the need for MPI code and separate HDF5/ASCII output.
* Do not hard-code number of accelerators per node, read SLURM env variables when available
* Allow file output frequency to be adjusted instead of writing recorded data only at the end of a simulation

**Pull request merged**